components/server/src/scm/scm-service.ts

```diff
@@ -0,0 +1,234 @@
+        if (repositoryService) {
+            log.info({ organizationId: teamId, userId: installer.id }, "Update prebuild installation for project.");
+            await repositoryService.installAutomatedPrebuilds(installer, cloneUrl);
+        }
+    }
+
+    /**
+     * `guessTokenScopes` requires the same permissions as `getToken`.
+     */
+    public async guessTokenScopes(
+        userId: string,
+        params: { host: string; repoUrl: string; gitCommand: string },
+    ): Promise<{ scopes?: string[]; message?: string }> {
+        const { host, repoUrl, gitCommand } = params;
+
+        const user = await this.userService.findUserById(userId, userId);
+
+        const provider = await this.authProviderService.findAuthProviderDescription(user, host);
+        if (!provider) {
+            throw new ApplicationError(ErrorCodes.NOT_FOUND, `Auth provider not found.`);
+        }
+
+        const { value: currentToken } = await this.getToken(userId, { host });
+        return await this.gitTokenScopeGuesser.guessGitTokenScopes(provider, {
+            host,
+            repoUrl,
+            gitCommand,
+            currentToken,
+        });
+    }
+
+    public async searchRepositories(userId: string, params: { searchString: string; limit?: number }) {
+        const user = await this.userService.findUserById(userId, userId);
+        const hosts = (await this.authProviderService.getAuthProviderDescriptions(user)).map((p) => p.host);
+
+        const limit: number = params.limit || 30;
+
+        const providerRepos = await Promise.all(
+            hosts.map(async (host): Promise<SuggestedRepositoryWithSorting[]> => {
+                try {
+                    const hostContext = this.hostContextProvider.get(host);
+                    const services = hostContext?.services;
+                    if (!services) {
+                        return [];
+                    }
+                    const repos = await services.repositoryProvider.searchRepos(user, params.searchString, limit);
+
+                    return repos.map((r) =>
+                        suggestionFromUserRepo({
+                            url: r.url.replace(/\.git$/, ""),
+                            repositoryName: r.name,
+                        }),
+                    );
+                } catch (error) {
+                    log.warn("Could not search repositories from host " + host, error);
+                }
+
+                return [];
+            }),
+        );
+
+        const sortedRepos = sortSuggestedRepositories(providerRepos.flat());
+
+        //return only the first 'limit' results
+        return sortedRepos.slice(0, limit).map(
+            (repo): SuggestedRepository => ({
+                url: repo.url,
+                repositoryName: repo.repositoryName,
+            }),
+        );
+    }
+
+    public async listSuggestedRepositories(
+        userId: string,
+        params: {
+            projectsPromise: Promise<Project[]>;
```

Comment:
⚠️ 

To avoid large refactoring and keep the current working solutions in Dashboard (New Workspace & New Project modals) untouched, this function still requires access to projects and workspaces. 

Passing in these dependencies as promises to resolve to something meaningful comes with the downside that the caller needs to parameterize the other services correctly. Alternatively, this could also be an async function to be passed as param 🤷🏻‍♂️  Would that make the 🌍 much better?

Reply:
Is there an issue to add dependency here to these services, a cycle?

Reply:
exactly. 

first of, it'd be a dependency cycle. but also, we've been talking about direction of dependencies, and it was agreed  on SCM not to depend on services like Orgs, Configs, Workspaces. 

not sure how to support the current Dashboard with a complete separation though. sometimes the idea is floating that the client is responsible to aggregate everything. we've tried that for the same use case (enriching a list of projects) and it was a nightmare with providing a defined order based on details fetched later/lazy, which we tried to compensate with more caching 🤯 

it seems like fast aggregation of data to support slick UIs is in conflict with pure API at times.

Reply:
yeah, I'm torn here. I also see a bit it like SCM service can have some low level functions which don't depend on other services, but then there are high level functions which can have a dependency. If you think like that then there is no really a conceptual cycle. If we wanted to make it more clear we will need to split SCM service to 2 services: one is low-level SCM access and another for aggregation. Passing params looks hacky.

cc @svenefftinge @geropl for an opinion

Reply:
on API level we should stick to one services, we don't want to have many one function services, mostly implementation question